packages/doenetml-worker-rust/lib-doenetml-core/src/core/props/traits/test_data_query_types.rs

```diff
@@ -0,0 +1,70 @@
+use itertools::Itertools;
+
+use crate::{components::ComponentVariantPropTypes, props::PropValueType};
+
+use super::TryFromDataQueryResults;
+
+/// This trait implements functions needed to test the types of a `DataQueryResults` object.
+/// It **should not be depended upon at runtime**, and is implemented only for testing type consistency.
+pub trait TestDataQueryTypes: TryFromDataQueryResults {
+    /// A list of tuples `(field_name, expected_type)` that are derived from the `RequiredData` struct's fields.
+    /// For example
+    /// ```ignore
+    /// struct RequiredData {
+    ///   a: PropView<prop_type::String>,
+    ///   b: PropView<prop_type::Number>,
+    /// }
+    /// ```
+    /// would result in the list `&[("a", Some(PropValueType::String)), ("b", Some(PropValueType::Number))]`.
+    const _DECLARED_DATA_QUERY_TYPES: &'static [(&'static str, Option<PropValueType>)];
+
+    const _STRUCT_NAME: &'static str;
+
+    /// Test that the types of the data queries match the types declared in the `RequiredData` struct.
+    /// This function should only be used as part of tests.
+    ///
+    /// It must be called with a type parameter of the related component. For example, if `RequiredData`
+    /// is for a prop on the `Text` component, this function is called
+    /// ```ignore
+    /// RequiredData::_test_data_query_types::<Text>();
+    /// ```
+    fn _test_data_query_types<T: ComponentVariantPropTypes + 'static>() -> anyhow::Result<()> {
```

Comment:
I'm wondering if there is a way to generalize this test to general props. I looked through all the current general props, and none of them have a `Prop` data query with `PropSpecifier::LocalIdx`. which seems to be the only reason one needs `ComponentVariantPropType` (other than for the error message, which obviously couldn't use a component name to identify the source of the error if it were a general prop).

Would it be hard to make it that if a `RequiredData` isn't given the `#[owning_component()]` attribute, then the test fails if one uses a `Prop` data query with `PropSpecifier::LocalIdx`, but otherwise works as normal (other than without the component name)? It'd be nice to have these checks when making general props.

Reply:
One could add a non-generic `_test_data_query_types` and write the tests manually for the general props.

Reply:
Sure, I could do that. But I can't think of a reason why I'd want anything customized. I want the same behavior as these, other than no support for `PropSpecifier::LocalIdx`. Or am I missing something?